File: trunk/java/SR-RBD/src/ddb/communication/UdpSenderImpl.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package ddb.communication;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;

import ddb.Config;
import ddb.Logger;
import ddb.msg.Message;

/** 
 * <!-- begin-UML-doc -->
 * <!-- end-UML-doc -->
 * @author Administrator
 * @generated "UML to Java (com.ibm.xtools.transform.uml2.java5.internal.UML2JavaTransform)"
 */
public class UdpSenderImpl implements UdpSender {

	/** 
	 * @generated "Singleton (com.ibm.xtools.patterns.content.gof.creational.singleton.SingletonPattern)"
	 */
	private final static String LOGGING_NAME = "UdpSenderImpl";
	private static UdpSenderImpl instance = new UdpSenderImpl();
	protected DatagramSocket socket;

	/** 
	 * @generated "Singleton (com.ibm.xtools.patterns.content.gof.creational.singleton.SingletonPattern)"
	 */
	private UdpSenderImpl() {
		// begin-user-code
		try {
			socket = new DatagramSocket();
		} catch (SocketException ex) {
			Logger.getInstance().log("SocketException in constructor: " + ex.getMessage(), 
					LOGGING_NAME, 
					Logger.Level.SEVERE);
		}
		// end-user-code
	}

	public static UdpSenderImpl getInstance() {
		return instance;
	}

	public synchronized void sendToAll(Message msg) {
		
		
		try {
			byte[] data = msg.Serialize();

			DatagramPacket packet = new DatagramPacket(data, data.length);
			packet.setAddress(InetAddress.getByName("255.255.255.255"));
			packet.setPort(Config.UdpPort());
			socket.send(packet);
			
		} catch (IOException ex) {
			Logger.getInstance().log("IOException: " + ex.getMessage(), LOGGING_NAME, Logger.Level.WARNING);
		}
	}
}

File: C#/RBD/RBD/Communication/UdpSenderImpl.cs
// +

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net;
using System.Net.Sockets;

using RBD.Msg;
namespace RBD
{
    class UdpSenderImpl : UdpSender
    {
        const String LOGGING_NAME = "UdpSenderImpl";
        static UdpSenderImpl instance = new UdpSenderImpl();
        protected Socket socket;

        IPEndPoint broadcast;

        private UdpSenderImpl()
        {
            try
            {
                broadcast = new IPEndPoint(IPAddress.Broadcast, Config.UdpPort());
                socket = new Socket(AddressFamily.InterNetwork,
                                    SocketType.Dgram, ProtocolType.Udp);
                socket.Connect(broadcast);
            }
            catch (SocketException ex)
            {
                Logger.getInstance().log("SocketException in constructor: " + ex.Message,
                        LOGGING_NAME,
                        Logger.Level.SEVERE);
            }
        }

        public static UdpSenderImpl getInstance()
        {
            return instance;
        }

        public void sendToAll(Message msg)
        {
            lock (this)
            {
                try
                {
                    byte[] data = msg.Serialize();
                    socket.Send(data);
                    Logger.getInstance().log("Sending: [" + msg.ToString() + "]", LOGGING_NAME, Logger.Level.INFO);
                }
                catch (SocketException ex)
                {
                    Logger.getInstance().log("SocketException: " + ex.Message, LOGGING_NAME, Logger.Level.WARNING);
                }
            }
        }
    }
}
